NetKet/Graph/hypercube.hpp

```diff
@@ -47,16 +50,216 @@ class Hypercube : public AbstractGraph {
+    symm_table_ = BuildSymmTable(L, ndim, pbc, nsites_);
+
+    colors_.reserve(nsites_);
+    for (auto const &edge : edges_) {
+      colors_.emplace(edge, 0);
+    }
+  }
+
+  // TODO(twesterhout): L is strictly speaking not needed, but then the logic
+  // becomes too complicated for my taste :)
+  // Also, the performance of this function will probably be pretty bad, by I
+  // don't think it matters much.
+  Hypercube(int const L, ColorMap colors)
+      : L_{L},
+        ndim_{},
+        pbc_{},
+        edges_{},
+        symm_table_{},
+        colors_{std::move(colors)} {
+    using std::begin;
+    using std::end;
+    if (colors_.empty()) {
+      throw InvalidInputError{
+          "Side length of the hypercube must be at least 1"};
+    }
+    // First of all, we extract the list of edges
+    edges_.reserve(colors_.size());
+    std::transform(
+        begin(colors_), end(colors_), std::back_inserter(edges_),
+        [](std::pair<AbstractGraph::Edge, int> const &x) { return x.first; });
+    // Verifies the edges and computes the number of sites
+    nsites_ = detail::CheckEdges(edges_);
+
+    // We know that L_^ndim_ == nsites_, so we can solve it for ndim_
+    auto count = L_;
+    ndim_ = 1;
+    while (count < nsites_) {
+      ++ndim_;
+      count *= L_;
+    }
+    if (count != nsites_) {
+      throw InvalidInputError{"Specified L and color map are incompatible"};
+    }
+
+    // For periodic boundary conditions, there are exactly ndim_ * nsites_ edges
+    // and for open bounrary conditions -- ndim_ * (nsites_ - nsites_/L).
+    if (static_cast<std::size_t>(ndim_ * nsites_) == edges_.size()) {
+      pbc_ = true;
+    } else if (static_cast<std::size_t>(ndim_ * (nsites_ - nsites_ / L)) ==
+               edges_.size()) {
+      pbc_ = false;
+    } else {
+      throw InvalidInputError{"Invalid color map"};
+    }
+
+    // Finally, we can check whether the edges we extracted from the color map
+    // make any sense.
+    auto const correct_edges = std::get<1>(BuildEdges(L_, ndim_, pbc_));
+    if (edges_.size() != correct_edges.size()) {
+      throw InvalidInputError{"Invalid color map"};
+    }
+    for (auto const &edge : correct_edges) {
+      // There is no anbiguity related to ordering: (i, j) vs (j, i), because
+      // CheckEdges asserts that i <= j for each (i, j).
+      if (!colors_.count(edge)) {
+        throw InvalidInputError{"Invalid color map"};
+      }
+    }
+
+    symm_table_ = BuildSymmTable(L_, ndim_, pbc_, nsites_);
+  }
+
+  int Nsites() const override { return nsites_; }
+
+  int Size() const override { return nsites_; }
+
+  int Length() const { return L_; }
+
+  int Ndim() const { return ndim_; }
+
+  const std::vector<Edge> &Edges() const noexcept { return edges_; }
+
+  std::vector<std::vector<int>> AdjacencyList() const override {
+    return detail::AdjacencyListFromEdges(Edges(), Nsites());
+  }
+
+  bool IsBipartite() const override { return L_ % 2 == 0; }
```

Comment:
What about `ndim=1, L=3`, i.e., the linear graph `O-O-O`? It is bipartite (since it is 2-colorable) but `L_ % 2 == 1`.

I think `IsBipartite` should always return `true` here. (Compare e.g. [Wikipedia](https://en.wikipedia.org/wiki/Lattice_graph#Properties).)

Reply:
Yes, I was hoping this would generate a few angry comments :) From the graph theory perspective the correct implementation is (I think):
```
return !pbc_ || L_ % 2 == 0;
```
which correctly handles the case you pointer out @femtobit.

The reason I wanted to generate some discussion around this point is the "let's change the sign of the off-diagonal elements in the Heisenberg Hamiltonian"-trick. It's currently done for bipartite graphs and I'm wondering, does it work okay (from the physics perspective) for @femtobit's test case? @gcarleo have you worked out the maths for it?